### DISCUSSION TITLE
Websocket client

### DISCUSSION THREAD

**Discussion by L4ngu0r** [tokio-rs/axum#653] [Q&A] [answered]
Hello,

I use `axum` to provide a proxy mecanism for websocket and thus use `tokio_tungstenite` to create a connection to other websocket server, as follow :

[ Client ] --- WS --> [axum ws | bridge | tungstenite ] --- WS ---> [ Server ]

I notice axum Message use methods [into_tungstenite](https://github.com/tokio-rs/axum/blob/dea36db400f27c025b646e5720b9a6784ea4db6e/axum/src/extract/ws.rs#L457) (and from variant) but it's private. As i want to dispatch websocket message to another server, I didn't find an easy way, to forward axum Message.

Also there is no method to detect clearly if message is a close event (as in warp: use `tungstenite::protocol::Message`)

Maybe I'm on the wrong way ? :)

**Comment by davidpdrsn** [comment, marked as answer]
Not quite sure what your question is 🤔

axum's use to tungstenite is private API so we won't expose any types from it. If we did that would requires us to ship a breaking release of axum is tungstenite does.

> Also there is no method to detect clearly if message is a close event (as in warp: use tungstenite::protocol::Message)

You can detect close messages by pattern matching and checking if it's `Message::Close`.

**Reply by L4ngu0r** [reply]
Yes thanks I always forgot about matching pattern :sweat_smile:
